auxtools/src/value_from.rs

```diff
@@ -0,0 +1,109 @@
+use crate::raw_types;
+use crate::runtime::Runtime;
+use crate::string;
+use crate::List;
+use crate::Value;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+impl From<f32> for Value {
+	fn from(num: f32) -> Self {
+		unsafe {
+			Value::new(
+				raw_types::values::ValueTag::Number,
+				raw_types::values::ValueData { number: num },
+			)
+		}
+	}
+}
+
+impl From<i32> for Value {
+	fn from(num: i32) -> Self {
+		unsafe {
+			Value::new(
+				raw_types::values::ValueTag::Number,
+				raw_types::values::ValueData { number: num as f32 },
+			)
+		}
+	}
+}
+
+impl From<u32> for Value {
+	fn from(num: u32) -> Self {
+		unsafe {
+			Value::new(
+				raw_types::values::ValueTag::Number,
+				raw_types::values::ValueData { number: num as f32 },
+			)
+		}
+	}
+}
+
+impl From<bool> for Value {
+	fn from(b: bool) -> Self {
+		unsafe {
+			Value::new(
+				raw_types::values::ValueTag::Number,
+				raw_types::values::ValueData {
+					number: if b { 1.0 } else { 0.0 },
+				},
+			)
+		}
+	}
+}
+
+impl From<&Value> for Value {
+	fn from(val: &Value) -> Self {
+		val.to_owned()
+	}
+}
+
+/* List-y helpers */
+
+// This is broken due to https://github.com/rust-lang/rust/issues/50133
+// The blanket implementation of TryFrom in core conflicts with -any- generics on a TryFrom trait
+//
+// impl<T: AsRef<str>> TryFrom<T> for Value {
+// 	type Error = Runtime;
+// 	fn try_from(value: T) -> Result<Self, Self::Error> {
+// 		Value::from_string(value.as_ref())
+// 	}
+// }
+
+// Specialized for ease-of-use due to the above not being possible
+impl<T: Into<Value> + Clone> TryFrom<&HashMap<String, T>> for Value {
+	type Error = Runtime;
+	fn try_from(hashmap: &HashMap<String, T>) -> Result<Self, Self::Error> {
+		let res = List::new();
+
+		for (k, v) in hashmap {
+			let string = string::StringRef::new(k)?;
+			res.set(string, v.clone())?;
+		}
+
+		Ok(res.into())
+	}
+}
+
+impl<A: Into<Value> + Clone, B: Into<Value> + Clone> From<&HashMap<A, B>> for Value {
+	fn from(hashmap: &HashMap<A, B>) -> Self {
+		let res = List::new();
+
+		for (k, v) in hashmap {
+			// Safety: This can't possibly fail with the trait bounds specifying both A and B are Into<Value>
+			res.set(k.clone(), v.clone()).unwrap();
```

Comment:
I'm pretty sure the possible runtimes from List::set are for things such as using null as a key. You might want to actually test that because it's a fairly rare code path.

Reply:
Yeah, this safety comment is definitely not correct with what bounds I actually have right now. I'm going to have to figure out something more rigid, or downgrade this to a TryFrom and just accept that some Value keys that are valid in a HashMap won't be for a BYOND list.

Reply:
Yeah, this safety comment is definitely not correct with what bounds I actually have right now. I'm going to have to figure out something more rigid, or downgrade this to a TryFrom and just accept that some Value keys that are valid in a HashMap won't be for a BYOND list.

Reply:
The latter seems more sane, tbh, let the user figure it out via BYOND screaming at them instead of trying to hardcode ValueTags that work as list indicies

Reply:
Even numerical indices should be able to cause a runtime by being out of bounds

Reply:
I don't think that's possible on a set action